### PR TITLE
nixos/static-web-server: create module which uses upstream systemd units

### PR DESCRIPTION
**Pull request by mac-chaffee**
###### Description of changes

This commit creates a nixos module for [static-web-server](https://static-web-server.net/). The module uses upstream systemd units to start static-web-server. It also includes options for configuring static-web-server. 

I also added a test which will install static-web-server with a custom port number, custom root dir, and custom config file, then curl it to see if it responds with a 200.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mac-chaffee** [comment]
@figsoda friendly ping, freshly rebased :)

File: nixos/modules/services/web-servers/static-web-server.nix [hunk 1]
```diff
@@ -0,0 +1,70 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
```
Comment (figsoda) [inline nixos/modules/services/web-servers/static-web-server.nix:3]: would be nice to remove this

File: pkgs/servers/static-web-server/default.nix [hunk 2]
```diff
@@ -24,6 +24,14 @@ rustPlatform.buildRustPackage rec {
     "--skip=handle_precondition"
   ];
 
+  # Need to copy in the systemd units for systemd.packages to discover them
+  postInstall = ''
+    install -Dm444 -t $out/lib/systemd/system/ systemd/static-web-server.service
+    install -Dm444 -t $out/lib/systemd/system/ systemd/static-web-server.socket
```
Comment (figsoda) [inline pkgs/servers/static-web-server/default.nix:30]:
```suggestion
    install -Dm444 -t $out/lib/systemd/system/ systemd/static-web-server.{service,socket}
```

File: nixos/tests/web-servers/static-web-server.nix [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import ../make-test-python.nix ({ pkgs, lib, ...} : {
+  name = "static-web-server";
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ mac-chaffee ];
```
Comment (figsoda) [inline nixos/tests/web-servers/static-web-server.nix:4]:
```suggestion
  meta = {
    maintainers = with lib.maintainers; [ mac-chaffee ];
```

File: nixos/tests/web-servers/static-web-server.nix [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import ../make-test-python.nix ({ pkgs, lib, ...} : {
```
Comment (figsoda) [inline nixos/tests/web-servers/static-web-server.nix:1]:
```suggestion
import ../make-test-python.nix ({ pkgs, lib, ... } : {
```

File: maintainers/maintainer-list.nix [hunk 5]
```diff
@@ -9769,6 +9769,11 @@
       fingerprint = "1147 43F1 E707 6F3E 6F4B  2C96 B9A8 B592 F126 F8E8";
     }];
   };
+  mac-chaffee = {
```
Comment (figsoda) [inline maintainers/maintainer-list.nix:9772]: please put this in a separate commit that precedes the other commit(s)
```
maintainers: add mac-chaffee
```

**Review by figsoda — COMMENTED** [review]
sorry about the delay, lgtm overall, just some nitpicks

File: nixos/tests/web-servers/static-web-server.nix [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+      enable = true;
+      listen = "[::]:8080";
+      root = "/tmp/";
+      configuration = {
+        general = { directory-listing = true; };
+      };
+    };
+  };
+
+  testScript = ''
```
Comment (figsoda) [inline nixos/tests/web-servers/static-web-server.nix:18]: would be nice if this also tests the files served, perhaps if we change `root` to a store path

**Comment by figsoda** [comment]
@ofborg build static-web-server static-web-server.tests

**Review by figsoda — APPROVED** [review]
Thank you!

**Comment by mac-chaffee** [comment]
Thanks for the review!

I also created a wiki page for using this module: https://nixos.wiki/wiki/Static_Web_Server

**Comment by joseluisq** [comment]
JFYI  The package and the module is now mentioned in the SWS docs. Thanks!
https://github.com/static-web-server/static-web-server/commit/eb2887f8dfa0d850d274eed8087eaa07722acebb
